src/tooltip.js

```diff
@@ -1,440 +1,327 @@
+                .append("strong")
+                .classed("x-value", true)
+                .html(headerFormatter(d.value));
+        }
 
-        };
+        var tbodyEnter = table.selectAll("tbody")
+            .data([d])
+            .enter().append("tbody");
+
+        var trowEnter = tbodyEnter.selectAll("tr")
+                .data(function(p) { return p.series})
+                .enter()
+                .append("tr")
+                .classed("highlight", function(p) { return p.highlight});
+
+        trowEnter.append("td")
+            .classed("legend-color-guide",true)
+            .append("div")
+            .style("background-color", function(p) { return p.color});
+
+        trowEnter.append("td")
+            .classed("key",true)
+            .classed("total",function(p) { return !!p.total})
+            .html(function(p, i) { return keyFormatter(p.key, i)});
+
+        trowEnter.append("td")
+            .classed("value",true)
+            .html(function(p, i) { return valueFormatter(p.value, i) });
+
+        trowEnter.selectAll("td").each(function(p) {
+            if (p.highlight) {
+                var opacityScale = d3.scale.linear().domain([0,1]).range(["#fff",p.color]);
+                var opacity = 0.6;
+                d3.select(this)
+                    .style("border-bottom-color", opacityScale(opacity))
+                    .style("border-top-color", opacityScale(opacity))
+                ;
+            }
+        });
+
+        var html = table.node().outerHTML;
+        if (d.footer !== undefined)
+            html += "<div class='footer'>" + d.footer + "</div>";
+        return html;
 
-        var dataSeriesExists = function(d) {
-            if (d && d.series) {
-                if (d.series instanceof Array) {
-                    return !!d.series.length;
-                }
-                // if object, it's okay just convert to array of the object
-                if (d.series instanceof Object) {
-                    d.series = [d.series];
-                    return true;
-                }
+    };
+
+    var dataSeriesExists = function(d) {
+        if (d && d.series) {
+            if (d.series instanceof Array) {
+                return !!d.series.length;
             }
-            return false;
-        };
+            // if object, it's okay just convert to array of the object
+            if (d.series instanceof Object) {
+                d.series = [d.series];
+                return true;
+            }
+        }
+        return false;
+    };
+
+    // Calculates the gravity offset of the tooltip. Parameter is position of tooltip
+    // relative to the viewport.
+    var calcGravityOffset = function(pos) {
+        var height = tooltip.node().offsetHeight,
+            width = tooltip.node().offsetWidth,
+            clientWidth = document.documentElement.clientWidth, // Don't want scrollbars.
+            clientHeight = document.documentElement.clientHeight, // Don't want scrollbars.
```

Comment:
http://stackoverflow.com/questions/1248081/get-the-browser-viewport-dimensions-with-javascript
http://andylangton.co.uk/blog/development/get-viewport-size-width-and-height-javascript

I used only clientWidth and clientHeight because innerWidth and innerHeight have wrong values on some mobile browsers.

This doesn't seem to be a problem.